### PR TITLE
Motion Matching: Multi-threaded feature extraction

### PR DESCRIPTION
**Pull request by amzn-jillich**
* Multi-threaded feature extraction using the job system as well as the task graph.
* The pose pool is not multi-thread safe. We added a pose pool parameter to the extraction context to control the pool usage manually.
* Rather than using the default pose pool for the given actor instance, we now use the given pose pool from the context when calculating the features.
* Features are now separated into batches and then extracted simultaneously, each job/task having its own pose pool.
* Added CVAR to enable/disable multi-threaded feature extraction.

Extracting features for 36k frames / 20mins motion database previously took ~6425ms and now takes around ~280ms using a 32-core Threadripper, 23x speedup.

Signed-off-by: Benjamin Jillich <jillich@amazon.com>

### CODE REVIEW

**Review by amzn-rhhong — APPROVED** [review]

File: Gems/MotionMatching/Code/Source/MotionMatchingData.h [hunk 1]
```diff
@@ -61,8 +61,14 @@ namespace EMotionFX::MotionMatching
         const AZStd::vector<Feature*>& GetFeaturesInKdTree() const { return m_featuresInKdTree; }
 
     protected:
+        //! Extract features from the motion database (multi-threaded).
         bool ExtractFeatures(ActorInstance* actorInstance, FrameDatabase* frameDatabase, size_t maxKdTreeDepth=20, size_t minFramesPerKdTreeNode=2000);
 
+        //! Extract features for a given range of frames and store the values in the feature matrix.
+        static void ExtractFeatureValuesRange(ActorInstance* actorInstance, FrameDatabase& frameDatabase, const FeatureSchema& featureSchema, FeatureMatrix& featureMatrix, size_t startFrame, size_t endFrame);
+
+        const size_t s_numFramesPerBatch = 1000; //!< Number of frames per task in the multi-threaded feature extraction routine.
```
Comment (jeremyong-az) [inline Gems/MotionMatching/Code/Source/MotionMatchingData.h:70]: Curious a bit how this number was derived, do you think this would need to vary per platform/hardware (possibly even be a runtime determination?)

**Review by jeremyong-az — APPROVED** [review]
Just had one comment/question, looks good!

File: Gems/MotionMatching/Code/Source/MotionMatchingData.h (hunk 1 above)
Comment (amzn-jillich) [reply Gems/MotionMatching/Code/Source/MotionMatchingData.h:70]: @jeremyong-az The two competing factors here are the overhead of the job/tasks (too many jobs and context switches) vs. a handful of allocations that we need to do per job/task. We need a unique pose pool per job/task as the pose pool class itself is not thread-safe (by design as we don't want to introduce locks in there). Poses are needed to calculate the joint velocities (as we're evaluating several frames of the motion and then integrating/averaging the velocities over a small time window) and a pose pool is needed to reuse the poses to avoid allocs. Though the initial set of poses needs to be allocated.

Long story short, I was playing with the numbers and 1k seemed to be close to the optimal point of the trade-off. The trade-off points varies along with the job overhead vs. allocation speed but as this is just about initialization and a one time thing vs. runtime, I am not sure if it is worth the time diving really deep into this?

File: Gems/MotionMatching/Code/Source/Feature.h [hunk 2]
```diff
@@ -76,6 +78,7 @@ namespace EMotionFX::MotionMatching
 
             size_t m_frameIndex = InvalidIndex;
             const Pose* m_framePose = nullptr; //! Pre-sampled pose for the given frame.
```
Comment (hultonha) [inline Gems/MotionMatching/Code/Source/Feature.h:80]: suggestion: Is it worth dropping `const` here as opposed to casting away `const` in `ExtractFeatureValuesRange`?

**Review by hultonha — COMMENTED** [review]
Looks really cool! Awesome speedup 👍

**Comment by amzn-jillich** [comment]
image
